src/kOS.Safe/Compilation/Opcode.cs

```diff
@@ -1401,32 +1401,64 @@ public override string ToString()
     {
         protected override string Name { get { return "return"; } }
         public override ByteCode Code { get { return ByteCode.RETURN; } }
+        
+        [MLField(0,true)]
+        public Int16 Depth { get; private set; } // Determines how many levels to popscope.
 
+        public override void PopulateFromMLFields(List<object> fields)
+        {
+            // Expect fields in the same order as the [MLField] properties of this class:
+            if (fields == null || fields.Count<1)
+                throw new Exception("Saved field in ML file for OpcodeCall seems to be missing.  Version mismatch?");
+            Depth = (Int16)fields[0];
+        }
+        
+        // Default constructor is needed for PopulateFromMLFields but shouldn't be used outside the KSM file handler:
+        private OpcodeReturn()
+        {
+        }
+        
+        /// <summary>
+        /// Make a return, telling it how many levels of the scope stack it should
+        /// be popping as it does so.  It compbines the behavior of a PopScope inside
+        /// itself, AFTER it reads and evaluates the thing atop the stack for return
+        /// purposes (that way it evals the top thing BEFORE it pops the scope and forgets
+        /// what variables exist).<br/
+        /// <br/>
+        /// Doing this:<br/>
+        ///   push $val<br/>
+        ///   return 2 deep<br/>
+        /// is the same as this:<br/>
+        ///   push $val<br/>
+        ///   eval<br/>
+        ///   popscope 2<br/>
+        ///   return 0 deep<br/>
+        /// <br/>
+        /// </summary>
+        /// <param name="popScopes"></param>
+        public OpcodeReturn(Int16 depth)
+        {
+            Depth = depth;
+        }
+        
         public override void Execute(ICpu cpu)
         {
-            // The only thing on the "above stack" now that is allowed to get in the way of
-            // finding the context record that tells us where to jump back to, are the potential
-            // closure scope frames that might have been pushed if this subroutine was
-            // called via a delegate reference.  Consume any of those that are in
-            // the way, then expect the context record.  Any other pattern encountered
-            // is proof the stack alignment got screwed up:
-            bool okay;
-            VariableScope peeked = cpu.PeekRaw(-1, out okay) as VariableScope;
-            while (okay && peeked != null && peeked.IsClosure)
-            {
-                cpu.PopAboveStack(1);
-                peeked = cpu.PeekRaw(-1, out okay) as VariableScope;
-            }
-            object shouldBeContextRecord = cpu.PopAboveStack(1);
-            if ( !(shouldBeContextRecord is SubroutineContext) )
-            {
-                // This should never happen with any user code:
-                throw new Exception( "kOS internal error: Stack misalignment detected when returning from routine.");
-            }
```

Comment:
This had to be moved further down because the SubroutineContext on the abovestack still has the scope frames in the way if the popscope is no longer happening before the return.  It has to wait until the internal "popscope" inside Return has happened before it tries this check, else it produces false positives.